### PR TITLE
add dns lookup time metric

### PR DESCRIPTION
**Pull request by yeganeahmadnejad**
add DNS metric to health exporter : health_http_dns_lookup_time_seconds

### CODE REVIEW

File: prober/http.go [hunk 1]
```diff
@@ -123,30 +141,76 @@ func (h *HTTP) calculateInterval() time.Duration {
...
+	var dnsDoneTime time.Time
+	var connectStartTime time.Time
+	var connectDoneTime time.Time
+	var tlsHandshakeStartTime time.Time
+	var tlsHandshakeDoneTime time.Time
+	httpTrace := &httptrace.ClientTrace{
+		GetConn: func(_ string) {
+		},
+		GotConn: func(info httptrace.GotConnInfo) {
+			log.Printf("conn was reused: %t", info.Reused)
```
Comment (m-yosefpor) [inline prober/http.go:156]:
```suggestion
			
```

File: prober/http.go [hunk 2]
```diff
@@ -123,30 +141,76 @@ func (h *HTTP) calculateInterval() time.Duration {
...
+
+	dnsLookupTime := dnsDoneTime.Sub(dnsStartTime)
+	dnsLookupTimeInMilliseconds := int(dnsLookupTime / time.Millisecond)
 
+	connectionHandshakeTime := connectDoneTime.Sub(connectStartTime)
+	connectionHandshakeTimeInMilliseconds := int(connectionHandshakeTime / time.Millisecond)
+
+	tlsHandshakeTime := tlsHandshakeDoneTime.Sub(tlsHandshakeStartTime)
+	tlsHandshakeTimeInMilliseconds := int(tlsHandshakeTime / time.Millisecond)
+	fmt.Println(responseTimeInMilliseconds, dnsLookupTimeInMilliseconds, connectionHandshakeTimeInMilliseconds, tlsHandshakeTimeInMilliseconds)
```
Comment (m-yosefpor) [inline prober/http.go:198]:
```suggestion

```

File: prober/http.go [hunk 3]
```diff
@@ -123,30 +141,76 @@ func (h *HTTP) calculateInterval() time.Duration {
...
-		req.Host = h.Host
+	var startTime time.Time
+	var finishTime time.Time
+	var dnsStartTime time.Time
+	var dnsDoneTime time.Time
+	var connectStartTime time.Time
+	var connectDoneTime time.Time
+	var tlsHandshakeStartTime time.Time
+	var tlsHandshakeDoneTime time.Time
+	httpTrace := &httptrace.ClientTrace{
```
Comment (alibo) [inline prober/http.go]: please extract the tracing codes into another method.

File: prober/http.go (hunk 1 above)
Comment (alibo) [inline prober/http.go]: please use klog instead.

Comment (alibo) [reply prober/http.go]: Also, please add useful metadata to the logs, e.g. URL, duration, etc.

File: prober/http.go [hunk 4]
```diff
@@ -123,30 +141,76 @@ func (h *HTTP) calculateInterval() time.Duration {
...
+		},
+		GotConn: func(info httptrace.GotConnInfo) {
+			log.Printf("conn was reused: %t", info.Reused)
+
+		},
+		DNSStart: func(_ httptrace.DNSStartInfo) {
+			dnsStartTime = time.Now()
+		},
+		DNSDone: func(_ httptrace.DNSDoneInfo) {
+			dnsDoneTime = time.Now()
```
Comment (alibo) [inline prober/http.go:152]: please check DNS errors and add required metrics

File: prober/http.go [hunk 5]
```diff
@@ -59,10 +70,11 @@ func NewHttp(name string, url string, rps float64, timeout time.Duration, tlsSki
...
 type HTTPResult struct {
-	ResponseTime float64
-	StatusCode   int
-	Error        error
-	ErrorType    string
+	ResponseTime  float64
+	StatusCode    int
+	Error         error
+	ErrorType     string
+	dnsLookupTime float64
```
Comment (alibo) [inline prober/http.go:75]: Please expose this variable

File: prober/http.go [hunk 6]
```diff
@@ -29,25 +31,34 @@ var (
...
+
 }
 
 func NewHttp(name string, url string, rps float64, timeout time.Duration, tlsSkipVerify bool, host string) HTTP {
 	client := &http.Client{
 		Timeout: timeout,
 	}
-
+	customTransport := http.DefaultTransport.(*http.Transport).Clone()
+	customTransport.DisableKeepAlives = true
```
Comment (alibo) [inline prober/http.go:55]: make this option configurable in the config file

File: prober/http.go [hunk 7]
```diff
@@ -29,25 +31,34 @@ var (
 		},
 		[]string{"name", "status_code", "result", "url"},
 	)
+	dnsLookupDurations = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "health_http_dns_lookup_time_seconds",
+			Help:    "The response time of dns lookup",
+			Buckets: []float64{0.001, 0.005, 0.01, 0.02, 0.03, 0.05, 0.075, 0.1, 0.2, 0.5, 0.75, 1, 1.5, 2, 2.5, 3, 4, 5},
+		},
+		[]string{"name", "status_code", "result", "url"},
```
Comment (alibo) [inline prober/http.go:38]: please add another label for dns_lookup_status or _error

File: prober/http.go (hunk 4 above)
Comment (alibo) [reply prober/http.go:152]: Also, log the errors with the required metadata

**Review by alibo — CHANGES_REQUESTED** [review]

**Comment by alibo** [comment]
Please, also update readme and sample config file.

File: prober/http.go (hunk 3 above)
Comment (m-yosefpor) [reply prober/http.go]: will be refactored in another PR

File: prober/http.go (hunk 1 above)
Comment (m-yosefpor) [reply prober/http.go]: this is removed

File: prober/http.go (hunk 6 above)
Comment (m-yosefpor) [reply prober/http.go:55]: will be fixed in anther PR.

File: prober/http.go (hunk 5 above)
Comment (m-yosefpor) [reply prober/http.go]: done

File: prober/http.go (hunk 6 above)
Comment (alibo) [reply prober/http.go:55]: ok, but it shouldn't go to production without this change.

File: prober/http.go (hunk 4 above)
Comment (m-yosefpor) [reply prober/http.go:155]: done

File: prober/http.go (hunk 7 above)
Comment (m-yosefpor) [reply prober/http.go]: done

File: prober/http.go (hunk 3 above)
Comment (alibo) [reply prober/http.go:150]: @m-yosefpor do you have any ETA on this?

**Pull request merged**